### PR TITLE
Feature/add elastic load balancer

### PR DESCRIPTION
**Pull request by sirococoa**
## Summary

- Support  elastic load balancer
- Update nifcloud-sdk-go

## Review
- [ ] Check Result

### 1. Create elastic load balancer in one arm mode

```
$ cat test.yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx
  labels:
    app: nginx
spec:
  replicas: 3
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
        - name: nginx
          image: nginx:alpine
          ports:
            - containerPort: 80
---
apiVersion: v1
kind: Service
metadata:
  name: nginx
  labels:
    app: nginx
  annotations:
    service.beta.kubernetes.io/nifcloud-load-balancer-type: elb
spec:
  type: LoadBalancer
  ports:
    - port: 80
      protocol: TCP
      targetPort: 80
  selector:
    app: nginx
$ kubectl apply -f test.yaml
```

- Elastic load balancer is created

image

image

- Security group rules are created

image

- Log output

```
$ kubectl -n kube-system logs daemonset.apps/nifcloud-cloud-controller-manager
I0118 07:19:39.552086       1 event.go:307] "Event occurred" object="default/nginx" fieldPath="" kind="Service" apiVersion="v1" type="Normal" reason="EnsuringLoadBalancer" message="Ensuring load balancer"
I0118 07:19:49.926643       1 nifcloud_elastic_load_balancer.go:66] Creating ElasticLoadBalancer "bf9481bb405f4f2" (80 -> 32249)
I0118 07:25:12.133763       1 event.go:307] "Event occurred" object="default/nginx" fieldPath="" kind="Service" apiVersion="v1" type="Normal" reason="EnsuredLoadBalancer" message="Ensured load balancer"
```

2. Add a port to an existing load balancer

```
$ cat test2.yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx
  labels:
    app: nginx
spec:
  replicas: 3
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
        - name: nginx
          image: nginx:alpine
          ports:
            - containerPort: 80
---
apiVersion: v1
kind: Service
metadata:
  name: nginx
  labels:
    app: nginx
  annotations:
    service.beta.kubernetes.io/nifcloud-load-balancer-type: elb
spec:
  type: LoadBalancer
  ports:
    - name: http
      port: 80
      protocol: TCP
      targetPort: 80
    - name: http2
      port: 8080
      protocol: TCP
      targetPort: 80
  selector:
    app: nginx
$ kubectl apply -f test2.yaml
```

- Elastic load balancer port is created

image

- Security group rules are created

image

- Log output

```
$ kubectl -n kube-system logs daemonset.apps/nifcloud-cloud-controller-manager
I0118 07:46:51.479391       1 event.go:307] "Event occurred" object="default/nginx" fieldPath="" kind="Service" apiVersion="v1" type="Normal" reason="EnsuringLoadBalancer" message="Ensuring load balancer"
I0118 07:46:53.671844       1 nifcloud_elastic_load_balancer.go:108] Registering ElasticLoadBalancer port "bf9481bb405f4f2" (8080 -> 30525)
I0118 07:49:13.743839       1 event.go:307] "Event occurred" object="default/nginx" fieldPath="" kind="Service" apiVersion="v1" type="Normal" reason="EnsuredLoadBalancer" message="Ensured load balancer"
```

3. Delete elastic load balancer

```
$ kubectl delete -f test2.yaml
```

- Log output

```
$ kubectl -n kube-system logs daemonset.apps/nifcloud-cloud-controller-manager
I0118 08:01:27.776096       1 event.go:307] "Event occurred" object="default/nginx" fieldPath="" kind="Service" apiVersion="v1" type="Normal" reason="DeletingLoadBalancer" message="Deleting load balancer"
I0118 08:01:28.383536       1 nifcloud_elastic_load_balancer.go:631] Deleting LoadBalancer "bf9481bb405f4f2" (80 -> 32249)
I0118 08:02:37.767129       1 nifcloud_elastic_load_balancer.go:631] Deleting LoadBalancer "bf9481bb405f4f2" (8080 -> 30525)
I0118 08:03:47.884602       1 event.go:307] "Event occurred" object="default/nginx" fieldPath="" kind="Service" apiVersion="v1" type="Normal" reason="DeletedLoadBalancer" message="Deleted load balancer"
```

- Elastic load balancers and security group rules created at 1. and 2. are deleted

4. Create elastic load balancer in two arm mode

```
$ cat test3.yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx
  labels:
    app: nginx
spec:
  replicas: 3
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
        - name: nginx
          image: nginx:alpine
          ports:
            - containerPort: 80
---
apiVersion: v1
kind: Service
metadata:
  name: nginx
  labels:
    app: nginx
  annotations:
    service.beta.kubernetes.io/nifcloud-load-balancer-type: elb
    service.beta.kubernetes.io/nifcloud-load-balancer-network-interface-1: net-COMMON_GLOBAL
    service.beta.kubernetes.io/nifcloud-load-balancer-network-interface-2: <private-network-id>
    service.beta.kubernetes.io/nifcloud-load-balancer-network-interface-2-ip-address: 192.168.2.149
    service.beta.kubernetes.io/nifcloud-load-balancer-network-interface-2-system-ip-address-1: 192.168.2.150
    service.beta.kubernetes.io/nifcloud-load-balancer-network-interface-2-system-ip-address-2: 192.168.2.151
    service.beta.kubernetes.io/nifcloud-load-balancer-vip-network: "1"
spec:
  type: LoadBalancer
  ports:
    - port: 80
      protocol: TCP
      targetPort: 80
  selector:
    app: nginx
$ kubectl apply -f test3.yaml
```
- Elastic load balancer is created

image

- Security group rules are created

image

- Log output

```
$ kubectl -n kube-system logs daemonset.apps/nifcloud-cloud-controller-manager
I0119 02:21:49.799116       1 event.go:307] "Event occurred" object="default/nginx" fieldPath="" kind="Service" apiVersion="v1" type="Normal" reason="EnsuringLoadBalancer" message="Ensuring load balancer"
I0119 02:21:54.270889       1 nifcloud_elastic_load_balancer.go:66] Creating ElasticLoadBalancer "ff42801fc1d9495" (80 -> 31864)
I0119 02:27:38.215040       1 event.go:307] "Event occurred" object="default/nginx" fieldPath="" kind="Service" apiVersion="v1" type="Normal" reason="EnsuredLoadBalancer" message="Ensured load balancer"
```

### CODE REVIEW

File: pkg/cloudprovider/providers/nifcloud/nifcloud_load_balancer.go [hunk 1]
```diff
@@ -14,14 +14,17 @@ import (
...
+	maxLoadBalancerNameLength   = 15
 	maxPortCountPerLoadBalancer = 3
 
 	// default health check parameter values
 	defaultHealthCheckInterval           = 10
 	defaultHealthCheckUnhealthyThreshold = 1
 	defaultHealthCheckTarget             = "TCP"
 
+	// default network interface
+	defaultNetworkInterface = "net-COMMON_GLOBAL"
```
[the remark this answers is not in the document]
Comment (sirococoa) [reply pkg/cloudprovider/providers/nifcloud/nifcloud_load_balancer.go]: I've corrected it as you suggested.

File: pkg/cloudprovider/providers/nifcloud/nifcloud_load_balancer.go [hunk 2]
```diff
@@ -55,10 +58,39 @@ const (
...
+	ServiceAnnotationLoadBalancerNetworkInterface2 = "service.beta.kubernetes.io/nifcloud-load-balancer-network-interface-2"
+
+	// ServiceAnnotationLoadBalancerNetworkInterface(1-2)IPAddress is the annotation that specify IPAdress of elastic load balancer
+	// Set IP address only when corresponding network interface is private
+	ServiceAnnotationLoadBalancerNetworkInterface1IPAddress = "service.beta.kubernetes.io/nifcloud-load-balancer-network-interface-1-ip-address"
+	ServiceAnnotationLoadBalancerNetworkInterface2IPAddress = "service.beta.kubernetes.io/nifcloud-load-balancer-network-interface-2-ip-address"
+
+	// ServiceAnnotationLoadBalancerNetworkInterface(1-2)SystemIPAddress(1-2) is the annotation that specify SystemIPAdress of elastic load balancer
+	// Set system IP address only when corresponding network interface is private
+	ServiceAnnotationLoadBalancerNetworkInterface1SystemIPAddress1 = "service.beta.kubernetes.io/nifcloud-load-balancer-network-interface-1-system-ip-address-1"
```
[the remark this answers is not in the document]
Comment (sirococoa) [reply pkg/cloudprovider/providers/nifcloud/nifcloud_load_balancer.go]: The annotation's format you suggested was easier to understand, so I adopted it.

In the reading process, fixed to simply separate them with a comma.

https://github.com/nifcloud/nifcloud-cloud-controller-manager/pull/18/commits/85bf351cad96ae175c864f581ca7f612134b08fc

File: pkg/cloudprovider/providers/nifcloud/nifcloud_elastic_load_balancer.go [hunk 3]
```diff
@@ -0,0 +1,700 @@
...
+const (
+	commonGlobalNetworkID  = "net-COMMON_GLOBAL"
+	commonPrivateNetworkID = "net-COMMON_PRIVATE"
+)
+
+func isElasticLoadBalancer(annotations map[string]string) bool {
+	return annotations[ServiceAnnotationLoadBalancerType] == "elb"
+}
+
+func isPrivateNetworkID(networkID string) bool {
```
[the remark this answers is not in the document]
Comment (sirococoa) [reply pkg/cloudprovider/providers/nifcloud/nifcloud_elastic_load_balancer.go]: The current function name is misleading, so I have fixed the one you suggested.

File: pkg/cloudprovider/providers/nifcloud/nifcloud_load_balancer.go [hunk 4]
```diff
@@ -55,10 +58,39 @@ const (
...
 	// See https://pfs.nifcloud.com/api/rest/ConfigureHealthCheck.htm
 	ServiceAnnotationLoadBalancerHCInterval = "service.beta.kubernetes.io/nifcloud-load-balancer-healthcheck-interval"
+
+	// ServiceAnnotationLoadBalancerType is the annotation that specify using load balancer type
+	// valid values are 'lb' or 'elb'
+	ServiceAnnotationLoadBalancerType = "service.beta.kubernetes.io/nifcloud-load-balancer-type"
+
+	// ServiceAnnotationLoadBalancerNetworkInterface(1-2) is the annotation that specify network interface of elastic load balancer
+	// net-COMMON_GLOBAL, net-COMMON_PRIVATE or network ID of private LAN
+	ServiceAnnotationLoadBalancerNetworkInterface1 = "service.beta.kubernetes.io/nifcloud-load-balancer-network-interface-1"
```
[the remark this answers is not in the document]
Comment (sirococoa) [reply pkg/cloudprovider/providers/nifcloud/nifcloud_load_balancer.go]: I've corrected it as you suggested.

File: pkg/cloudprovider/providers/nifcloud/nifcloud_elastic_load_balancer.go [hunk 5]
```diff
@@ -0,0 +1,700 @@
...
+	loadBalancerName := c.GetLoadBalancerName(ctx, clusterName, service)
+
+	// describe load balancer
+	loadBalancers, err := c.client.DescribeElasticLoadBalancers(ctx, loadBalancerName)
+	if err != nil {
+		return nil, false, err
+	}
+
+	if len(loadBalancers) == 0 {
+		return nil, false, fmt.Errorf("not found load balancer: %q", loadBalancerName)
```
[the remark this answers is not in the document]
Comment (sirococoa) [reply pkg/cloudprovider/providers/nifcloud/nifcloud_elastic_load_balancer.go]: I've corrected it as you suggested.

File: pkg/cloudprovider/providers/nifcloud/nifcloud_client.go [hunk 6]
```diff
@@ -55,6 +62,48 @@ type Filter struct {
...
+	Protocol                      string
+	NetworkVolume                 int32
+	BalancingType                 int32
+	BalancingTargets              []Instance
+	LoadBalancerPort              int32
+	InstancePort                  int32
+	HealthCheckTarget             string
+	HealthCheckInterval           int32
+	HealthCheckUnhealthyThreshold int32
+	NetworkInterface              []NetworkInterface
```
[the remark this answers is not in the document]
Comment (sirococoa) [reply pkg/cloudprovider/providers/nifcloud/nifcloud_client.go]: I've corrected it as you suggested.

File: pkg/cloudprovider/providers/nifcloud/nifcloud_elastic_load_balancer.go [hunk 7]
```diff
@@ -0,0 +1,700 @@
...
+
+		if !found {
+			diff = append(diff, x)
+		}
+	}
+
+	return diff
+}
+
+func toElasticLoadBalancerStatus(vip string) *v1.LoadBalancerStatus {
```
[the remark this answers is not in the document]
Comment (sirococoa) [reply pkg/cloudprovider/providers/nifcloud/nifcloud_elastic_load_balancer.go]: I've corrected it as you suggested.

File: pkg/cloudprovider/providers/nifcloud/nifcloud_load_balancer.go [hunk 8]
```diff
@@ -55,10 +58,39 @@ const (
 	// ServiceAnnotationLoadBalancerHCInterval is the annotation that specify interval seconds for health check
 	// See https://pfs.nifcloud.com/api/rest/ConfigureHealthCheck.htm
 	ServiceAnnotationLoadBalancerHCInterval = "service.beta.kubernetes.io/nifcloud-load-balancer-healthcheck-interval"
+
+	// ServiceAnnotationLoadBalancerType is the annotation that specify using load balancer type
+	// valid values are 'lb' or 'elb'
+	ServiceAnnotationLoadBalancerType = "service.beta.kubernetes.io/nifcloud-load-balancer-type"
```
[the remark this answers is not in the document]
Comment (sirococoa) [reply pkg/cloudprovider/providers/nifcloud/nifcloud_load_balancer.go:64]: The process of load balancer will be separated into other functions and files.

I would like to handle it at the same time.

File: pkg/cloudprovider/providers/nifcloud/nifcloud_elastic_load_balancer.go [hunk 9]
```diff
@@ -0,0 +1,700 @@
...
+				systemIPAdresses = append(systemIPAdresses, systemIPAddress2)
+			}
+			if len(systemIPAdresses) == 2 {
+				networkInterface.SystemIpAddresses = systemIPAdresses
+			} else {
+				return nil, fmt.Errorf("system ip address require two value")
+			}
+		}
+
+		if vipNetwork, ok := annotations[ServiceAnnotationLoadBalancerVipNetwork]; ok {
```
[the remark this answers is not in the document]
Comment (sirococoa) [reply pkg/cloudprovider/providers/nifcloud/nifcloud_elastic_load_balancer.go:337]: I think that there is a lack of validation not only in IsVipNetwork.

I will add validation for IsVipNetwork at the same time as I add validation for other annotations.

[the remark this answers is not in the document]
Comment (aokumasan) [reply pkg/cloudprovider/providers/nifcloud/nifcloud_elastic_load_balancer.go:337]: OK!

File: pkg/cloudprovider/providers/nifcloud/nifcloud_load_balancer.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (aokumasan) [reply pkg/cloudprovider/providers/nifcloud/nifcloud_load_balancer.go:64]: OK. Thank you!!

**Comment by aokumasan** [comment]
Thank you! It works!! 🚀

**Pull request merged**